pythonwhat/local.py

```diff
@@ -66,3 +53,46 @@ def run_code(process, code):
         raw_output = ""
         error = str(e)
     return raw_output, error
+
+
+def run_single_process(pec, code, pid=None):
+    process = StubProcess(init_code=pec, pid=pid)
+    raw_stu_output, error = run_code(process, code)
+    return process, raw_stu_output, error
+
+
+def run_exercise(pec, sol_code, stu_code, pid=None, sol_wd=None, stu_wd=None):
+    with ChDir(stu_wd or os.getcwd()):
+        stu_process, raw_stu_output, error = run_single_process(pec, stu_code, pid=pid)
+
+    with ChDir(sol_wd or os.getcwd()):
+        sol_process, _, _ = run_single_process(pec, sol_code, pid=pid)
+
+    return sol_process, stu_process, raw_stu_output, error
+
+
+# TODO:
+#  - test has_equal_value: has_equal_value (also with name and/or expr_code)
+#    - check_object currently only works on root, this introduces a new 'root' for that purpose
+#  - converge with pythonbackend, look at scalabackend
+#    - support running with arbitrary wd + path + flags (now only wd)
+#      e.g. `python -m project.run
+#      allow setting env vars? e.g. PYTHONPATH, could help running more complex setup
+def run(state, relative_workin_directory=""):
+    sol_wd = Path(os.getcwd(), "solution", relative_workin_directory)
+    os.makedirs(str(sol_wd), exist_ok=True)
+    stu_wd = Path(os.getcwd(), relative_workin_directory)
+    sol_process, stu_process, raw_stu_output, error = run_exercise(
```

Comment:
When we use this in combination with `check_files`, it only works when `solution_code` is given, right? We're probably going to run into issues when there are import statements from other files in the same directory. E.g. if `script.py` has this as a solution:

```python
from .cool_functions import super_cool_function

super_cool_function("THIS IS AWESOME")
```

It will be even weirder if the user asked for the solution. In this case, it would actually work.

Reply:
`solution_code` is needed if checks depend on the solution. An example where this is not the case is `has_equal_value` with an `override`: https://github.com/datacamp/pythonwhat/blob/f7b5ea60c0e76cf06dc59acda71adfd590ad1e81/tests/test_check_files.py#L119-L130

Reply:
You are right about the limitations of importing other local scripts.
This is something we could probably tackle when we work on automatically providing the solution code to the SCT.